File: apps/desktop/src/services/llm.ts
/**
 * LLM provider adapters.
 *
 * SECURITY: API keys come exclusively from the caller (via useLLMConfig).
 * Nothing is hardcoded. Keys never leave the browser except to the specific
 * provider endpoint the user explicitly configured.
 */

import type { LLMConfig, LLMProvider } from '../hooks/useLLMConfig';

export interface LLMCallOptions {
  maxTokens?: number;
  temperature?: number;
}

/* ── per-provider call implementations ──────────────────────────── */

async function callAnthropic(
  cfg: LLMConfig,
  prompt: string,
  system: string,
  opts: LLMCallOptions,
): Promise<string> {
  const res = await fetch('https://api.anthropic.com/v1/messages', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'x-api-key': cfg.apiKey,
      'anthropic-version': '2023-06-01',
      // Required header to allow direct browser access to the Anthropic API
      'anthropic-dangerous-direct-browser-access': 'true',
    },
    body: JSON.stringify({
      model: cfg.model || 'claude-sonnet-4-6',
      max_tokens: opts.maxTokens ?? 150,
      ...(system ? { system } : {}),
      messages: [{ role: 'user', content: prompt }],
    }),
  });
  if (!res.ok) {
    const err = await res.json().catch(() => ({})) as Record<string, unknown>;
    const msg = (err as Record<string, Record<string, string>>)?.error?.message ?? `HTTP ${res.status}`;
    throw new Error(msg);
  }
  const data = await res.json() as { content: Array<{ text: string }> };
  return data.content[0].text.trim();
}

async function callOpenAI(
  cfg: LLMConfig,
  prompt: string,
  system: string,
  opts: LLMCallOptions,
): Promise<string> {
  const base = cfg.baseUrl.replace(/\/$/, '') || 'https://api.openai.com';
  const res = await fetch(`${base}/v1/chat/completions`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      Authorization: `Bearer ${cfg.apiKey}`,
    },
    body: JSON.stringify({
      model: cfg.model || 'gpt-4o-mini',
      max_tokens: opts.maxTokens ?? 150,
      temperature: opts.temperature ?? 0.8,
      messages: [
        ...(system ? [{ role: 'system', content: system }] : []),
        { role: 'user', content: prompt },
      ],
    }),
  });
  if (!res.ok) {
    const err = await res.json().catch(() => ({})) as Record<string, unknown>;
    const msg = (err as Record<string, Record<string, string>>)?.error?.message ?? `HTTP ${res.status}`;
    throw new Error(msg);
  }
  const data = await res.json() as { choices: Array<{ message: { content: string } }> };
  return data.choices[0].message.content.trim();
}

async function callGemini(
  cfg: LLMConfig,
  prompt: string,
  system: string,
  opts: LLMCallOptions,
): Promise<string> {
  const model = cfg.model || 'gemini-2.0-flash';
  const url = `https://generativelanguage.googleapis.com/v1beta/models/${model}:generateContent?key=${cfg.apiKey}`;
  const combined = system ? `${system}\n\n${prompt}` : prompt;
  const res = await fetch(url, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      contents: [{ parts: [{ text: combined }] }],
      generationConfig: {
        maxOutputTokens: opts.maxTokens ?? 150,
        temperature: opts.temperature ?? 0.8,
      },
    }),
  });
  if (!res.ok) {
    const err = await res.json().catch(() => ({})) as Record<string, unknown>;
    const msg = String((err as Record<string, Record<string, string>>)?.error?.message ?? `HTTP ${res.status}`);
    throw new Error(msg);
  }
  const data = await res.json() as {
    candidates: Array<{ content: { parts: Array<{ text: string }> } }>;
  };
  return data.candidates[0].content.parts[0].text.trim();
}

async function callOllama(
  cfg: LLMConfig,
  prompt: string,
  system: string,
  opts: LLMCallOptions,
): Promise<string> {
  const base = cfg.baseUrl.replace(/\/$/, '') || 'http://localhost:11434';
  const res = await fetch(`${base}/api/chat`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      model: cfg.model || 'llama3',
      stream: false,
      options: { num_predict: opts.maxTokens ?? 150, temperature: opts.temperature ?? 0.8 },
      messages: [
        ...(system ? [{ role: 'system', content: system }] : []),
        { role: 'user', content: prompt },
      ],
    }),
  });
  if (!res.ok) throw new Error(`Ollama HTTP ${res.status} — is it running?`);
  const data = await res.json() as { message: { content: string } };
  return data.message.content.trim();
}

/* ── dispatcher ─────────────────────────────────────────────────── */

const ADAPTERS: Record<LLMProvider, typeof callAnthropic> = {
  anthropic: callAnthropic,
  openai: callOpenAI,
  gemini: callGemini,
  ollama: callOllama,
};

export async function callLLM(
  config: LLMConfig,
  prompt: string,
  system = '',
  maxTokens?: number,
  temperature?: number,
): Promise<string> {
  const fn = ADAPTERS[config.provider];
  if (!fn) throw new Error(`Unknown provider: ${config.provider}`);
  return fn(config, prompt, system, { maxTokens, temperature });
}

/** Wraps callLLM with a timeout so boot sequences never stall. */
export async function callLLMSafe(
  config: LLMConfig,
  prompt: string,
  system = '',
  maxTokens = 80,
  timeoutMs = 6000,
): Promise<string | null> {
  try {
    const race = await Promise.race([
      callLLM(config, prompt, system, maxTokens, 0.85),
      new Promise<never>((_, rej) => setTimeout(() => rej(new Error('timeout')), timeoutMs)),
    ]);
    return race;
  } catch {
    return null;
  }
}

File: apps/orchestrator/app/services/llm.py
from __future__ import annotations

import httpx

VOICE_SYSTEM_PROMPT = (
    'You are Robo, a helpful AI voice assistant. '
    'You answer only general knowledge questions. '
    'You have no access to the user\'s personal data or real-time information — '
    'those come exclusively from connected agents. '
    'Reply in 1–3 short sentences. No markdown, no bullet points — plain spoken language only.'
)


class LLMService:
    """Thin async wrapper over multiple LLM providers. Returns plain text or None on failure."""

    async def complete(
        self,
        user_message: str,
        llm_config: dict,
        system_prompt: str = VOICE_SYSTEM_PROMPT,
        max_tokens: int = 250,
        temperature: float = 0.7,
    ) -> str | None:
        provider = (llm_config.get('provider') or 'openai').lower()
        api_key  = (llm_config.get('api_key')  or '').strip()
        model    = (llm_config.get('model')     or '').strip()
        base_url = (llm_config.get('base_url')  or '').strip().rstrip('/')

        try:
            if provider == 'openai':
                return await self._openai_compat(
                    system_prompt, user_message, api_key,
                    model or 'gpt-4o-mini',
                    base_url or 'https://api.openai.com',
                    max_tokens, temperature,
                )
            if provider == 'ollama':
                return await self._openai_compat(
                    system_prompt, user_message, '',
                    model or 'llama3',
                    base_url or 'http://localhost:11434',
                    max_tokens, temperature,
                )
            if provider == 'anthropic':
                return await self._anthropic(
                    system_prompt, user_message,
                    api_key, model or 'claude-haiku-4-5-20251001', max_tokens, temperature,
                )
            if provider == 'gemini':
                return await self._gemini(
                    system_prompt, user_message,
                    api_key, model or 'gemini-2.0-flash', max_tokens, temperature,
                )
        except Exception:
            return None
        return None

    # ── OpenAI / Ollama (OpenAI-compatible) ──────────────────────────────────

    async def _openai_compat(
        self,
        system: str,
        user: str,
        api_key: str,
        model: str,
        base_url: str,
        max_tokens: int,
        temperature: float = 0.7,
    ) -> str | None:
        headers: dict = {'Content-Type': 'application/json'}
        if api_key:
            headers['Authorization'] = f'Bearer {api_key}'
        async with httpx.AsyncClient(timeout=20.0) as client:
            r = await client.post(
                f'{base_url}/v1/chat/completions',
                headers=headers,
                json={
                    'model': model,
                    'messages': [
                        {'role': 'system', 'content': system},
                        {'role': 'user',   'content': user},
                    ],
                    'max_tokens': max_tokens,
                    'temperature': temperature,
                },
            )
            if r.status_code == 200:
                return r.json()['choices'][0]['message']['content'].strip() or None
        return None

    # ── Anthropic ─────────────────────────────────────────────────────────────

    async def _anthropic(
        self, system: str, user: str, api_key: str, model: str, max_tokens: int,
        temperature: float = 0.7,
    ) -> str | None:
        async with httpx.AsyncClient(timeout=20.0) as client:
            r = await client.post(
                'https://api.anthropic.com/v1/messages',
                headers={
                    'x-api-key':          api_key,
                    'anthropic-version':  '2023-06-01',
                    'content-type':       'application/json',
                },
                json={
                    'model':       model,
                    'system':      system,
                    'messages':    [{'role': 'user', 'content': user}],
                    'max_tokens':  max_tokens,
                    'temperature': temperature,
                },
            )
            if r.status_code == 200:
                return r.json()['content'][0]['text'].strip() or None
        return None

    # ── Google Gemini ─────────────────────────────────────────────────────────

    async def _gemini(
        self, system: str, user: str, api_key: str, model: str, max_tokens: int,
        temperature: float = 0.7,
    ) -> str | None:
        combined = f'{system}\n\nUser: {user}'
        async with httpx.AsyncClient(timeout=20.0) as client:
            r = await client.post(
                f'https://generativelanguage.googleapis.com/v1beta/models/{model}:generateContent',
                params={'key': api_key},
                json={
                    'contents':         [{'parts': [{'text': combined}]}],
                    'generationConfig': {'maxOutputTokens': max_tokens, 'temperature': temperature},
                },
            )
            if r.status_code == 200:
                text = r.json()['candidates'][0]['content']['parts'][0]['text'].strip()
                return text or None
        return None


# Module-level singleton shared across agents
llm_service = LLMService()
